otter/test/util/test_zk.py

```diff
@@ -0,0 +1,194 @@
+        if version != -1:
+            current_version = self.nodes[path][1]
+            if current_version != version:
+                return fail(BadVersionError(
+                    "When operating on {}, version {} was specified by "
+                    "version {} was found".format(path, version,
+                                                  current_version)))
+
+    def set(self, path, new_value, version=-1):
+        """Set the content of a node."""
+        check = self._check_version(path, version)
+        if check is not None:
+            return check
+        current_version = self.nodes[path][1]
+        new_stat = ZNodeStatStub(version=current_version + 1)
+        self.nodes[path] = (new_value, new_stat.version)
+        return succeed(new_stat)
+
+    def delete(self, path, version=-1):
+        """Delete a node."""
+        check = self._check_version(path, version)
+        if check is not None:
+            return check
+        del self.nodes[path]
+        return succeed(None)
+
+
+class CreateOrSetTests(SynchronousTestCase):
+    """Tests for :func:`create_or_set`."""
+    def setUp(self):
+        self.model = ZKCrudModel()
+
+    def _cos(self, path, content):
+        eff = Effect(CreateOrSet(path=path, content=content))
+        performer = partial(perform_create_or_set, self.model)
+        dispatcher = TypeDispatcher({CreateOrSet: performer})
+        return perform(dispatcher, eff)
+
+    def test_create(self):
+        """Creates a node when it doesn't exist."""
+        d = self._cos('/foo', 'bar')
+        self.assertEqual(self.successResultOf(d), '/foo')
+        self.assertEqual(self.model.nodes, {'/foo': ('bar', 0)})
+
+    def test_update(self):
+        """Uses `set` to update the node when it does exist."""
+        self.model.create('/foo', 'initial', makepath=True)
+        d = self._cos('/foo', 'bar')
+        self.assertEqual(self.successResultOf(d), '/foo')
+        self.assertEqual(self.model.nodes, {'/foo': ('bar', 1)})
+
+    def test_node_disappears_during_update(self):
+        """
+        If `set` can't find the node (because it was unexpectedly deleted
+        between the `create` and `set` calls), creation will be retried.
+        """
+        def hacked_set(path, value):
+            self.model.delete('/foo')
+            del self.model.set  # Only let this behavior run once
+            return self.model.set(path, value)
+        self.model.set = hacked_set
+
+        self.model.create('/foo', 'initial', makepath=True)
+        d = self._cos('/foo', 'bar')
+        self.assertEqual(self.successResultOf(d), '/foo')
+        # It must be at version 0 because it's a creation, whereas normally if
+        # the node were being updated it'd be at version 1.
+        self.assertEqual(self.model.nodes, {'/foo': ('bar', 0)})
+
+    def test_loop_limit(self):
+        """
+        performing a :obj:`CreateOrSet` will only try to loop between creating
+        and setting 100 times.
```

Comment:
Non-blocking: The test is not really checking the limit of 100. You could change the doc to "performing createorset will not go in infinite loop and eventually give up with CreateOrSetLimitReachedError"